### PR TITLE
Adding OIDCClient class and tests

### PR DESCRIPTION
**Pull request by cduongt**

### CODE REVIEW

File: .rubocop.yml [hunk 1]
```diff
@@ -11,9 +11,13 @@ Style/Documentation:
 
 Metrics/MethodLength:
   Max: 15
+  Exclude:
```
Comment (amarthadan) [inline .rubocop.yml]: You know that it kinda loses sense to use RuboCop if you will exclude from the check the only file in this pull request that actually does something, right?

File: .rubocop.yml [hunk 2]
```diff
@@ -11,9 +11,13 @@ Style/Documentation:
...
+    - 'lib/keyrod/oidc_client.rb'
 
 Metrics/AbcSize:
   Max: 20
+  Exclude:
```
Comment (amarthadan) [inline .rubocop.yml]: Ditto.

File: config/config.yml [hunk 3]
```diff
@@ -1,11 +1,14 @@
 keyrod:
+  debug: false # Debug mode
+  ca-dir: # CA directory
+  verify-ssl: true # Verify SSL certificate
   fedcloud:
-    site: http://localhost/ # EGI FedCloud Site to authenticate against
```
Comment (amarthadan) [inline config/config.yml]: Why did you remove this?

File: config/config.yml [hunk 4]
```diff
@@ -1,11 +1,14 @@
...
+    site: # EGI FedCloud Site to authenticate against
     access_token: # Access token
     group: # ID of the group to join
     interactive-fallback: false # Fallback to interactive mode if group is not set
   oidc:
-    site: http://localhost/token # OIDC site to authenticate refresh token against
```
Comment (amarthadan) [inline config/config.yml]: And this?

File: lib/keyrod.rb [hunk 5]
```diff
@@ -1,6 +1,8 @@
 module Keyrod
   autoload :CLI, 'keyrod/cli'
   autoload :Settings, 'keyrod/settings'
+  autoload :OIDCClient, 'keyrod/oidc_client'
+  autoload :Errors, 'keyrod/error'
```
Comment (amarthadan) [inline lib/keyrod.rb]: Module is called `Errors` and file is called `error.rb`. This goes against the best practices.

File: lib/keyrod/cli.rb [hunk 6]
```diff
@@ -51,6 +66,22 @@ class CLI < Thor
            type: :string,
            desc: 'OIDC client secret',
            aliases: '-t'
-    def token; end
+    def token
+      merge_config options
```
Comment (amarthadan) [inline lib/keyrod/cli.rb:71]: You're missing a call that would actually do something. But I guess this is on purpose for now.

File: lib/keyrod/oidc_client.rb [hunk 7]
```diff
@@ -0,0 +1,34 @@
+require 'faraday'
+require 'json'
+
+module Keyrod
+  class OIDCClient
+    def initialize
+      @oidc_endpoint = Keyrod::Settings['oidc-site']
```
Comment (amarthadan) [inline lib/keyrod/oidc_client.rb]: When accessing `Settings` use symbols instead of strings.

File: lib/keyrod/oidc_client.rb [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+      @refresh_token = Keyrod::Settings['refresh-token']
+      @client_id = Keyrod::Settings['client-id']
+      @client_secret = Keyrod::Settings['client-secret']
+    end
+
+    def access_token
+      ssl = { verify: Keyrod::Settings['verify-ssl'] }
+      ssl['ca_path'] = Keyrod::Settings['ca-dir'] if Keyrod::Settings['ca-dir']
+      conn = Faraday.new(@oidc_endpoint, ssl: ssl) do |faraday|
```
Comment (amarthadan) [inline lib/keyrod/oidc_client.rb]: When accessing object fields use accessor instead of the direct access.

File: lib/keyrod/oidc_client.rb [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+      end
+
+      response = conn.post do |req|
+        req.params['grant_type'] = 'refresh_token'
+        req.params['client_id'] = @client_id
+        req.params['client_secret'] = @client_secret
+        req.params['refresh_token'] = @refresh_token
+        req.params['scope'] = 'openid profile'
+      end
+      raise Keyrod::Errors::ResponseError, "Response from OIDC server was #{response.status}" if response.status != 200
```
Comment (amarthadan) [inline lib/keyrod/oidc_client.rb]: Faraday response has a method `success?` that is supposed to be used to check the response status.

File: lib/keyrod/oidc_client.rb [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+      @refresh_token = Keyrod::Settings['refresh-token']
+      @client_id = Keyrod::Settings['client-id']
+      @client_secret = Keyrod::Settings['client-secret']
+    end
+
+    def access_token
+      ssl = { verify: Keyrod::Settings['verify-ssl'] }
+      ssl['ca_path'] = Keyrod::Settings['ca-dir'] if Keyrod::Settings['ca-dir']
+      conn = Faraday.new(@oidc_endpoint, ssl: ssl) do |faraday|
+        faraday.request  :url_encoded
```
Comment (amarthadan) [inline lib/keyrod/oidc_client.rb]: Are these Faraday settings really necessary? Aren't they the default?

File: lib/keyrod/oidc_client.rb [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+        faraday.response :logger if Keyrod::Settings['debug']
```
Comment (amarthadan) [inline lib/keyrod/oidc_client.rb]: You should use logger such as `Yell` and not just log on `STDOUT`. This goes for the whole application not just the Faraday requests.

File: lib/keyrod/oidc_client.rb [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+        faraday.adapter  Faraday.default_adapter
+      end
+
+      response = conn.post do |req|
+        req.params['grant_type'] = 'refresh_token'
```
Comment (amarthadan) [inline lib/keyrod/oidc_client.rb]: You should use symbols instead of strings when setting request params.

File: lib/keyrod/oidc_client.rb [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+    def access_token
+      ssl = { verify: Keyrod::Settings['verify-ssl'] }
+      ssl['ca_path'] = Keyrod::Settings['ca-dir'] if Keyrod::Settings['ca-dir']
+      conn = Faraday.new(@oidc_endpoint, ssl: ssl) do |faraday|
+        faraday.request  :url_encoded
+        faraday.response :logger if Keyrod::Settings['debug']
+        faraday.adapter  Faraday.default_adapter
+      end
+
+      response = conn.post do |req|
```
Comment (amarthadan) [inline lib/keyrod/oidc_client.rb]: Are you relly supposed to use POST and then encode parameters in the URL? It can be done just it seems a bit weird to me.

File: lib/keyrod/oidc_client.rb [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+        req.params['grant_type'] = 'refresh_token'
+        req.params['client_id'] = @client_id
+        req.params['client_secret'] = @client_secret
+        req.params['refresh_token'] = @refresh_token
+        req.params['scope'] = 'openid profile'
+      end
+      raise Keyrod::Errors::ResponseError, "Response from OIDC server was #{response.status}" if response.status != 200
+
+      JSON.parse(response.body)['access_token']
```
Comment (amarthadan) [inline lib/keyrod/oidc_client.rb]: You should call `JSON.parse` with option `symbolize_names: true` and then access the result hash via symbols.

File: spec/keyrod/oidc_client_spec.rb [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+    Keyrod::Settings['client-secret'] = '3HcNV88ex48UKYPWCI1OwresZAxMjh1Pdf'
+  end
+
+  describe '#new' do
+    it 'returns instance of OIDCClient' do
+      is_expected.to be_instance_of described_class
+    end
+  end
+
+  describe '#access_token' do
```
Comment (amarthadan) [inline spec/keyrod/oidc_client_spec.rb]: This should be called `.access_token`. `#` is used for class methods and `.` for object methods.

File: spec/keyrod/oidc_client_spec.rb [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+  subject(:oidc_client) { described_class.new }
+
+  before do
+    Keyrod::Settings['oidc-site'] = 'https://aai-dev.egi.eu/oidc/token'
+    Keyrod::Settings['refresh-token'] = 'iOiI0ZDQ5MzUxMC00MDJhLTQ3NDQtYjE4Yi'
+    Keyrod::Settings['client-id'] = 'keyrod'
+    Keyrod::Settings['client-secret'] = '3HcNV88ex48UKYPWCI1OwresZAxMjh1Pdf'
+  end
+
+  describe '#new' do
```
Comment (amarthadan) [inline spec/keyrod/oidc_client_spec.rb]: Maybe you should also test whether those object variables are really set as you expect (judging from the constructor implementation).

File: spec/keyrod/oidc_client_spec.rb (hunk 15 above)
Comment (amarthadan) [inline spec/keyrod/oidc_client_spec.rb]: You should also test for invalid execution.

File: spec/spec_helper.rb [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+
+WebMock.disable_net_connect!(allow_localhost: true)
+
+RSpec.configure do |config|
+  config.color = true
+  config.tty = true
+  config.order = 'random'
+
+  config.before do
+    stub_request(:post, 'https://aai-dev.egi.eu/oidc/token?client_id=keyrod&client_secret=3HcNV88ex48UKYPWCI1OwresZAxMjh1Pdf&grant_type=refresh_token&refresh_token=iOiI0ZDQ5MzUxMC00MDJhLTQ3NDQtYjE4Yi&scope=openid%20profile')
```
Comment (amarthadan) [inline spec/spec_helper.rb]: You should do the stubbing in the test class you are actually testing not in the general configuration.

File: spec/spec_helper.rb [hunk 18]
```diff
@@ -0,0 +1,21 @@
+require 'webmock/rspec'
+require 'keyrod'
+
+WebMock.disable_net_connect!(allow_localhost: true)
+
+RSpec.configure do |config|
+  config.color = true
```
Comment (amarthadan) [inline spec/spec_helper.rb]: This configuration option is deprecated according to the configuration.

File: spec/spec_helper.rb [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+  config.tty = true
```
Comment (amarthadan) [inline spec/spec_helper.rb]: What does this option do?

**Review by amarthadan — CHANGES_REQUESTED** [review]

File: .rubocop.yml (hunk 1 above)
Comment (sustr4) [reply .rubocop.yml]: Well. it excludes it from just one metric. But you are right, this should not be necessary since this is new code. @cduongt Do we really have a good enough reason to exclude the check rather than satisfy it?

File: .rubocop.yml (hunk 2 above)
Comment (sustr4) [reply .rubocop.yml]: Ditto :smile:

File: .rubocop.yml (hunk 1 above)
Comment (amarthadan) [reply .rubocop.yml]: `MethodLength` and `AbcSize` are one of those main metrices that keeps code clean, divided into multiple smaller methods and that way testable. And as you pointed out it's a new code.

Comment (sustr4) [reply .rubocop.yml]: :+1:

File: lib/keyrod/oidc_client.rb (hunk 8 above)
Comment (cduongt) [reply lib/keyrod/oidc_client.rb]: I'm not sure what you mean. Hash fields should be set by hash[:key], or am I wrong?

File: lib/keyrod/oidc_client.rb (hunk 13 above)
Comment (cduongt) [reply lib/keyrod/oidc_client.rb]: According to http://openid.net/specs/openid-connect-core-1_0.html#RefreshingAccessToken , the request should be POST and should use form serialization

File: lib/keyrod/oidc_client.rb (hunk 8 above)
Comment (amarthadan) [reply lib/keyrod/oidc_client.rb]: I'm not talking about hash but about object attributes. When accessing object attributes (or fields), in this case `@oidc_endpoint`, you should use [accessors](https://ruby-doc.org/docs/ruby-doc-bundle/UsersGuide/rg/accessors.html) (also known as getters in other languages).

File: lib/keyrod/oidc_client.rb (hunk 13 above)
Comment (amarthadan) [reply lib/keyrod/oidc_client.rb]: :+1:

File: .rubocop.yml (hunk 2 above)
Comment (cduongt) [reply .rubocop.yml]: For some reason this metric counts string interpolation aswell, which I need for logging. I will keep this file excluded for this metric for now.

Comment (amarthadan) [reply .rubocop.yml]: No, you won't. It's a new class with simple code and simple functionality. If its ABC size is too big it's because it's poorly written. If the ABC size is too big dividing method into multiple smaller methods usually helps.

**Comment by amarthadan** [comment]
I'll let it go this time but commit message with text _refactored code a bit_ and _fixed issues_ means nothing to me. In the future please address each issue (and feature) in separate commit. That way I can easily see if you addressed everything I required in the review.

File: spec/keyrod/oidc_client_spec.rb [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+      .to_return(status: 400, body: '', headers: {})
+  end
+
+  describe '#new' do
+    it 'returns instance of OIDCClient' do
+      is_expected.to be_instance_of described_class
+    end
+
+    it 'sets oidc-site' do
+      expect(oidc_client.oidc_site).to be_instance_of(String)
```
Comment (amarthadan) [inline spec/keyrod/oidc_client_spec.rb]: I guess you can test if those attributes are Strings but I don't really see the point. What you should test for is whether those Strings contain what you expect them to, in this case strings from the settings.

File: lib/keyrod/oidc_client.rb [hunk 21]
```diff
@@ -0,0 +1,44 @@
+require 'faraday'
+require 'json'
+
+module Keyrod
+  class OIDCClient
+    attr_reader :oidc_site, :refresh_token, :client_id, :client_secret, :post_params, :ssl
```
Comment (amarthadan) [inline lib/keyrod/oidc_client.rb]: Do you really need `post_params` and `ssl` object attributes? I would suggest renaming `set_params` and `set_ssl` methods to `prepare_params` and `prepare_ssl` or just `params` and `ssl`, make them return the values and call these methods directly from `access_token` method without storing the values in the object attributes. What do you think?

File: spec/keyrod/oidc_client_spec.rb [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+      expect(oidc_client.client_id).to be_instance_of(String)
+    end
+
+    it 'sets client-secret' do
+      expect(oidc_client.client_secret).to be_instance_of(String)
+    end
+  end
+
+  describe '.access_token' do
+    it 'returns access token' do
```
Comment (amarthadan) [inline spec/keyrod/oidc_client_spec.rb]: This test should go under the `with successful run` context.

File: spec/keyrod/oidc_client_spec.rb [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+      expect(oidc_client.access_token).to be_instance_of(String)
+    end
+
+    context 'with successful run' do
+      it 'doesnt raise any errors' do
```
Comment (amarthadan) [inline spec/keyrod/oidc_client_spec.rb]: Typo: `doesnt` -> `doesn't`.

**Review by amarthadan — CHANGES_REQUESTED** [review]

File: lib/keyrod/oidc_client.rb (hunk 21 above)
Comment (cduongt) [reply lib/keyrod/oidc_client.rb]: I tried that in the first place, however it raised AbcSize metric over 20.

Comment (amarthadan) [reply lib/keyrod/oidc_client.rb]: That is not possible because it won't change `access_token` method in any way.

File: lib/keyrod/oidc_client.rb [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+    def params
+      { grant_type: 'refresh_token',
+        client_id: client_id,
+        client_secret: client_secret,
+        refresh_token: refresh_token,
+        scope: 'openid profile' }
+    end
+
+    def ssl
+      ssl = { verify: Keyrod::Settings[:'verify-ssl'] }
```
Comment (amarthadan) [inline lib/keyrod/oidc_client.rb]: Please don't use the same name for a variable as for a method. It may lead to some unexpected errors later. (I know it will work, it's more of a style issue).

**Review by amarthadan — CHANGES_REQUESTED** [review]

**Review by amarthadan — APPROVED** [review]
